File: packages/node/src/adapters/replicate.ts
import Replicate from "replicate";
import { ProviderAdapter } from "../core/provider.js";
import {
  GenerateInput,
  GenerateOutput,
  LipsyncGenerateInput,
  LipsyncGenerateOutput,
  ListModelsParams,
  ModelMetadata,
  Provider,
  ReplicateProviderConfig,
  StreamChunk,
  VideoGenerateInput,
  VideoGenerateOutput,
} from "../core/types.js";
import { AiKitError } from "../core/errors.js";
import { ErrorKind } from "../core/types.js";

const VIDEO_MODELS: ModelMetadata[] = [
  {
    id: "kwaivgi/kling-v2.5-turbo-pro",
    displayName: "Kling v2.5 Turbo Pro",
    provider: Provider.Replicate,
    family: "i2v",
    capabilities: {
      text: true,
      vision: true,
      image: false,
      video: true,
      tool_use: false,
      structured_output: false,
      reasoning: false,
    },
    videoPrices: {
      per_second_usd: 0.07,
    },
  },
  {
    id: "kwaivgi/kling-v2.6",
    displayName: "Kling v2.6",
    provider: Provider.Replicate,
    family: "i2v",
    capabilities: {
      text: true,
      vision: true,
      image: false,
      video: true,
      tool_use: false,
      structured_output: false,
      reasoning: false,
      audio_out: true,
    },
    videoPrices: {
      per_second_usd: 0.07,
      per_second_usd_with_audio: 0.14,
    },
  },
];

const LIPSYNC_MODELS: ModelMetadata[] = [
  {
    id: "kwaivgi/kling-lip-sync",
    displayName: "Kling Lip Sync",
    provider: Provider.Replicate,
    family: "lipsync",
    capabilities: {
      text: true,
      vision: false,
      image: false,
      video: true,
      video_in: true,
      tool_use: false,
      structured_output: false,
      reasoning: false,
      audio_in: true,
    },
    videoPrices: {
      per_second_usd: 0.014,
    },
  },
  {
    id: "bytedance/latentsync",
    displayName: "LatentSync",
    provider: Provider.Replicate,
    family: "lipsync",
    capabilities: {
      text: false,
      vision: false,
      image: false,
      video: true,
      video_in: true,
      tool_use: false,
      structured_output: false,
      reasoning: false,
      audio_in: true,
    },
    videoPrices: {
      per_request_usd: 0.072,
    },
  },
  {
    id: "pixverse/lipsync",
    displayName: "PixVerse Lipsync",
    provider: Provider.Replicate,
    family: "lipsync",
    capabilities: {
      text: false,
      vision: false,
      image: false,
      video: true,
      video_in: true,
      tool_use: false,
      structured_output: false,
      reasoning: false,
      audio_in: true,
    },
    videoPrices: {
      per_second_usd: 0.04,
    },
  },
];

async function downloadAsBase64(url: string): Promise<string> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to download video: ${response.statusText}`);
  }
  const buffer = await response.arrayBuffer();
  return Buffer.from(buffer).toString("base64");
}

export class ReplicateAdapter implements ProviderAdapter {
  readonly provider = Provider.Replicate;
  private readonly client: Replicate;

  constructor(config: ReplicateProviderConfig) {
    this.client = new Replicate({
      auth: config.apiKey,
    });
  }

  async listModels(_params?: ListModelsParams): Promise<ModelMetadata[]> {
    return [...VIDEO_MODELS, ...LIPSYNC_MODELS];
  }

  async generate(_input: GenerateInput): Promise<GenerateOutput> {
    throw new AiKitError({
      kind: ErrorKind.Unsupported,
      message: "Replicate text generation is not supported",
      provider: this.provider,
    });
  }

  async *streamGenerate(_input: GenerateInput): AsyncIterable<StreamChunk> {
    throw new AiKitError({
      kind: ErrorKind.Unsupported,
      message: "Replicate text generation is not supported",
      provider: this.provider,
    });
  }

  async generateVideo(input: VideoGenerateInput): Promise<VideoGenerateOutput> {
    const modelInput: Record<string, unknown> = {
      prompt: input.prompt,
    };

    if (input.startImage) {
      modelInput.start_image = input.startImage;
    } else if (input.inputImages?.length) {
      const img = input.inputImages[0];
      if (img.url) {
        modelInput.start_image = img.url;
      } else if (img.base64) {
        const prefix = img.mediaType ? `data:${img.mediaType};base64,` : "data:image/png;base64,";
        modelInput.start_image = prefix + img.base64;
      }
    }

    if (input.duration !== undefined) {
      modelInput.duration = input.duration;
    }

    if (input.aspectRatio) {
      modelInput.aspect_ratio = input.aspectRatio;
    }

    if (input.negativePrompt) {
      modelInput.negative_prompt = input.negativePrompt;
    }

    if (input.generateAudio !== undefined) {
      modelInput.generate_audio = input.generateAudio;
    }

    if (input.parameters) {
      Object.assign(modelInput, input.parameters);
    }

    const output = await this.client.run(input.model as `${string}/${string}`, {
      input: modelInput,
    });

    let videoUrl: string;
    if (typeof output === "string") {
      videoUrl = output;
    } else if (Array.isArray(output) && output.length > 0) {
      videoUrl = String(output[0]);
    } else if (output && typeof output === "object" && "url" in output) {
      videoUrl = String((output as { url: string }).url);
    } else {
      throw new AiKitError({
        kind: ErrorKind.Unknown,
        message: "Unexpected output format from Replicate",
        provider: this.provider,
      });
    }

    const data = await downloadAsBase64(videoUrl);

    return {
      mime: "video/mp4",
      data,
      raw: output,
    };
  }

  async generateLipsync(input: LipsyncGenerateInput): Promise<LipsyncGenerateOutput> {
    const modelInput: Record<string, unknown> = {};

    // Video input
    if (input.videoUrl) {
      modelInput.video_url = input.videoUrl;
      modelInput.video = input.videoUrl;
    } else if (input.videoBase64) {
      const prefix = "data:video/mp4;base64,";
      modelInput.video_url = prefix + input.videoBase64;
      modelInput.video = prefix + input.videoBase64;
    }

    // Audio input
    if (input.audioUrl) {
      modelInput.audio_file = input.audioUrl;
      modelInput.audio = input.audioUrl;
    } else if (input.audioBase64) {
      const prefix = "data:audio/mp3;base64,";
      modelInput.audio_file = prefix + input.audioBase64;
      modelInput.audio = prefix + input.audioBase64;
    }

    // Text-to-speech options (Kling Lip Sync only)
    if (input.text) {
      modelInput.text = input.text;
    }
    if (input.voiceId) {
      modelInput.voice_id = input.voiceId;
    }
    if (input.voiceSpeed !== undefined) {
      modelInput.voice_speed = input.voiceSpeed;
    }

    if (input.parameters) {
      Object.assign(modelInput, input.parameters);
    }

    const output = await this.client.run(input.model as `${string}/${string}`, {
      input: modelInput,
    });

    let videoUrl: string;
    if (typeof output === "string") {
      videoUrl = output;
    } else if (Array.isArray(output) && output.length > 0) {
      videoUrl = String(output[0]);
    } else if (output && typeof output === "object" && "url" in output) {
      videoUrl = String((output as { url: string }).url);
    } else {
      throw new AiKitError({
        kind: ErrorKind.Unknown,
        message: "Unexpected output format from Replicate",
        provider: this.provider,
      });
    }

    const data = await downloadAsBase64(videoUrl);

    return {
      mime: "video/mp4",
      data,
      raw: output,
    };
  }
}

File: packages/python-inference/src/ai_kit/providers/replicate.py
from __future__ import annotations

import base64
import io
import os
from dataclasses import dataclass
from typing import Any, List, Optional, Sequence

from ..catalog import load_catalog_models
from ..errors import ErrorKind, KitErrorPayload, AiKitError
from ..types import ImageGenerateInput, ImageGenerateOutput, ImageInput, ModelCapabilities, ModelMetadata
from ..clients.replicate_client import ReplicateClient


_NANO_BANANA_PREFIX = "google/nano-banana"


@dataclass
class ReplicateConfig:
    api_key: str = ""
    api_keys: Optional[List[str]] = None


class ReplicateAdapter:
    def __init__(self, config: ReplicateConfig) -> None:
        self.config = config
        if config.api_key:
            os.environ["REPLICATE_API_TOKEN"] = config.api_key
        self._client = ReplicateClient()

    def list_models(self) -> List[ModelMetadata]:
        models = [model for model in load_catalog_models() if model.provider == "replicate"]
        if models:
            return models
        return [
            ModelMetadata(
                id=_NANO_BANANA_PREFIX,
                displayName=_NANO_BANANA_PREFIX,
                provider="replicate",
                family="image-edit",
                capabilities=ModelCapabilities(
                    text=True,
                    vision=True,
                    image=True,
                    tool_use=False,
                    structured_output=False,
                    reasoning=False,
                ),
            )
        ]

    def generate_image(self, input: ImageGenerateInput) -> ImageGenerateOutput:
        if not input.model.startswith(_NANO_BANANA_PREFIX):
            raise AiKitError(
                KitErrorPayload(
                    kind=ErrorKind.UNSUPPORTED,
                    message=f"Replicate adapter only supports {_NANO_BANANA_PREFIX}",
                    provider="replicate",
                )
            )
        image_inputs = _coerce_image_inputs(input.inputImages)
        parameters = dict(input.parameters or {})
        output_format = parameters.pop("output_format", "png")
        aspect_ratio = parameters.pop("aspect_ratio", None)
        result = self._client.nano_banana(
            model=input.model,
            prompt=input.prompt,
            image_input=image_inputs,
            aspect_ratio=aspect_ratio,
            output_format=output_format,
            parameters=parameters or None,
        )
        data = base64.b64encode(result).decode("ascii")
        mime = f"image/{output_format}" if output_format else "image/png"
        return ImageGenerateOutput(mime=mime, data=data)

    def generate(self, input):
        raise AiKitError(
            KitErrorPayload(
                kind=ErrorKind.UNSUPPORTED,
                message="Replicate text generation is not supported",
                provider="replicate",
            )
        )

    def generate_mesh(self, input):
        raise AiKitError(
            KitErrorPayload(
                kind=ErrorKind.UNSUPPORTED,
                message="Replicate mesh generation is not supported",
                provider="replicate",
            )
        )

    def generate_speech(self, input):
        raise AiKitError(
            KitErrorPayload(
                kind=ErrorKind.UNSUPPORTED,
                message="Replicate speech generation is not supported",
                provider="replicate",
            )
        )


def _coerce_image_inputs(raw: Optional[Sequence[Any]]) -> Optional[List[Any]]:
    if not raw:
        return None
    inputs: List[Any] = []
    for entry in raw:
        image = _normalize_image_entry(entry)
        if image is None:
            continue
        inputs.append(image)
    return inputs or None


def _normalize_image_entry(entry: Any) -> Any:
    if isinstance(entry, ImageInput):
        return _image_input_to_payload(entry)
    if isinstance(entry, dict):
        return _image_input_to_payload(ImageInput(**entry))
    if isinstance(entry, str):
        if entry.startswith("data:"):
            return io.BytesIO(_decode_base64(entry))
        return entry
    return entry


def _image_input_to_payload(entry: ImageInput) -> Any:
    if entry.url:
        return entry.url
    if entry.base64:
        data = _decode_base64(entry.base64)
        return io.BytesIO(data)
    return None


def _decode_base64(value: str) -> bytes:
    if value.startswith("data:"):
        payload = value.split(",", 1)[1]
        return base64.b64decode(payload)
    return base64.b64decode(value)
